File: tests/runtime/variants/runner.go
package runner

import (
	"context"
	"fmt"

	wrpc "wrpc.io/go"

	"driver/runner/test/variants/to_test"
)

func Run(ctx context.Context, c wrpc.Invoker) error {
	assertEqual(*must(to_test.RoundtripOption(ctx, c, ptr[float32](1.0))), 1)
	assertNil(must(to_test.RoundtripOption(ctx, c, nil)))
	assertEqual(*must(to_test.RoundtripOption(ctx, c, ptr[float32](2.0))), 2)

	assertEqual(*must(to_test.RoundtripResult(ctx, c, wrpc.Ok[float32, uint32](2))).Ok, 2.0)
	assertEqual(*must(to_test.RoundtripResult(ctx, c, wrpc.Ok[float32, uint32](4))).Ok, 4.0)
	assertEqual(*must(to_test.RoundtripResult(ctx, c, wrpc.Err[uint32, float32](5.3))).Err, 5)

	assertEqual(must(to_test.RoundtripEnum(ctx, c, to_test.E1_A)), to_test.E1_A)
	assertEqual(must(to_test.RoundtripEnum(ctx, c, to_test.E1_B)), to_test.E1_B)

	assertEqual(must(to_test.InvertBool(ctx, c, true)), false)
	assertEqual(must(to_test.InvertBool(ctx, c, false)), true)

	{
		a1, a2, a3, a4, a5, a6 := must6(to_test.VariantCasts(ctx, c, &to_test.Casts{
			V0: to_test.NewC1A(1),
			V1: to_test.NewC2A(2),
			V2: to_test.NewC3A(3),
			V3: to_test.NewC4A(4),
			V4: to_test.NewC5A(5),
			V5: to_test.NewC6A(6.0),
		}))
		assertVariant(a1.GetA())
		assertVariant(a2.GetA())
		assertVariant(a3.GetA())
		assertVariant(a4.GetA())
		assertVariant(a5.GetA())
		assertVariant(a6.GetA())
	}

	{
		a1, a2, a3, a4, a5, a6 := must6(to_test.VariantCasts(ctx, c, &to_test.Casts{
			V0: to_test.NewC1B(1),
			V1: to_test.NewC2B(2.0),
			V2: to_test.NewC3B(3.0),
			V3: to_test.NewC4B(4.0),
			V4: to_test.NewC5B(5.0),
			V5: to_test.NewC6B(6.0),
		}))
		assertVariant(a1.GetB())
		assertVariant(a2.GetB())
		assertVariant(a3.GetB())
		assertVariant(a4.GetB())
		assertVariant(a5.GetB())
		assertVariant(a6.GetB())
	}

	{
		a1, a2, a3, a4 := must4(to_test.VariantZeros(ctx, c, &to_test.Zeros{
			V0: to_test.NewZ1A(1),
			V1: to_test.NewZ2A(2),
			V2: to_test.NewZ3A(3.0),
			V3: to_test.NewZ4A(4.0),
		}))
		assertVariant(a1.GetA())
		assertVariant(a2.GetA())
		assertVariant(a3.GetA())
		assertVariant(a4.GetA())
	}

	{
		a1, a2, a3, a4 := must4(to_test.VariantZeros(ctx, c, &to_test.Zeros{
			V0: to_test.NewZ1B(),
			V1: to_test.NewZ2B(),
			V2: to_test.NewZ3B(),
			V3: to_test.NewZ4B(),
		}))
		assertEqual(a1.Discriminant(), to_test.Z1B)
		assertEqual(a2.Discriminant(), to_test.Z2B)
		assertEqual(a3.Discriminant(), to_test.Z3B)
		assertEqual(a4.Discriminant(), to_test.Z4B)
	}

	must0(to_test.VariantTypedefs(ctx, c, nil, false, wrpc.Err[uint32, struct{}](struct{}{})))

	{
		a, b, cc := must3(to_test.VariantEnums(ctx, c, true, wrpc.Ok[struct{}, struct{}](struct{}{}), to_test.MyErrno_Success))
		assertEqual(a, true)
		assert(b.Ok != nil)
		assertEqual(cc, to_test.MyErrno_Success)
	}
	return nil
}

func ptr[T any](v T) *T {
	return &v
}

func must[T any](v T, err error) T {
	if err != nil {
		panic(err)
	}
	return v
}

func must0(err error) {
	if err != nil {
		panic(err)
	}
}

func must3[A, B, C any](a A, b B, c C, err error) (A, B, C) {
	if err != nil {
		panic(err)
	}
	return a, b, c
}

func must4[A, B, C, D any](a A, b B, c C, d D, err error) (A, B, C, D) {
	if err != nil {
		panic(err)
	}
	return a, b, c, d
}

func must6[A, B, C, D, E, F any](a A, b B, c C, d D, e E, f F, err error) (A, B, C, D, E, F) {
	if err != nil {
		panic(err)
	}
	return a, b, c, d, e, f
}

func assertVariant[T any](_ T, ok bool) {
	if !ok {
		panic("unexpected variant discriminant")
	}
}

func assertNil[T any](v *T) {
	if v != nil {
		panic(fmt.Sprintf("expected nil, got %v", *v))
	}
}

func assertEqual[T comparable](a T, b T) {
	if a != b {
		panic(fmt.Sprintf("%v not equal to %v", a, b))
	}
}

func assert(v bool) {
	if !v {
		panic("assertion failed")
	}
}

File: tests/runtime/variants/runner.rs
use crate::runner::test::variants::to_test::*;

pub async fn run(
    wrpc: &impl ::wit_bindgen_wrpc::wrpc_transport::Invoke<Context = ()>,
) -> ::wit_bindgen_wrpc::anyhow::Result<()> {
    assert_eq!(roundtrip_option(wrpc, (), Some(1.0)).await?, Some(1));
    assert_eq!(roundtrip_option(wrpc, (), None).await?, None);
    assert_eq!(roundtrip_option(wrpc, (), Some(2.0)).await?, Some(2));
    assert_eq!(roundtrip_result(wrpc, (), &Ok(2)).await?, Ok(2.0));
    assert_eq!(roundtrip_result(wrpc, (), &Ok(4)).await?, Ok(4.0));
    assert_eq!(roundtrip_result(wrpc, (), &Err(5.3)).await?, Err(5));

    assert_eq!(roundtrip_enum(wrpc, (), E1::A).await?, E1::A);
    assert_eq!(roundtrip_enum(wrpc, (), E1::B).await?, E1::B);

    assert_eq!(invert_bool(wrpc, (), true).await?, false);
    assert_eq!(invert_bool(wrpc, (), false).await?, true);

    let (a1, a2, a3, a4, a5, a6) = variant_casts(
        wrpc,
        (),
        (C1::A(1), C2::A(2), C3::A(3), C4::A(4), C5::A(5), C6::A(6.0)),
    )
    .await?;
    assert!(matches!(a1, C1::A(1)));
    assert!(matches!(a2, C2::A(2)));
    assert!(matches!(a3, C3::A(3)));
    assert!(matches!(a4, C4::A(4)));
    assert!(matches!(a5, C5::A(5)));
    assert!(matches!(a6, C6::A(b) if b == 6.0));

    let (a1, a2, a3, a4, a5, a6) = variant_casts(
        wrpc,
        (),
        (
            C1::B(1),
            C2::B(2.0),
            C3::B(3.0),
            C4::B(4.0),
            C5::B(5.0),
            C6::B(6.0),
        ),
    )
    .await?;
    assert!(matches!(a1, C1::B(1)));
    assert!(matches!(a2, C2::B(b) if b == 2.0));
    assert!(matches!(a3, C3::B(b) if b == 3.0));
    assert!(matches!(a4, C4::B(b) if b == 4.0));
    assert!(matches!(a5, C5::B(b) if b == 5.0));
    assert!(matches!(a6, C6::B(b) if b == 6.0));

    let (a1, a2, a3, a4) =
        variant_zeros(wrpc, (), (Z1::A(1), Z2::A(2), Z3::A(3.0), Z4::A(4.0))).await?;
    assert!(matches!(a1, Z1::A(1)));
    assert!(matches!(a2, Z2::A(2)));
    assert!(matches!(a3, Z3::A(b) if b == 3.0));
    assert!(matches!(a4, Z4::A(b) if b == 4.0));

    let (a1, a2, a3, a4) = variant_zeros(wrpc, (), (Z1::B, Z2::B, Z3::B, Z4::B)).await?;
    assert!(matches!(a1, Z1::B));
    assert!(matches!(a2, Z2::B));
    assert!(matches!(a3, Z3::B));
    assert!(matches!(a4, Z4::B));

    variant_typedefs(wrpc, (), None, false, &Err(())).await?;

    assert_eq!(
        variant_enums(wrpc, (), true, &Ok(()), MyErrno::Success).await?,
        (true, Ok(()), MyErrno::Success)
    );
    Ok(())
}
